### ISSUE TITLE
mallocMC managing two or more contexts

### ISSUE THREAD

**Issue by anshumang** [ComputationalRadiationPhysics/mallocMC#99]
@slizzered @ax3l Thank you for developing this library and providing the sources :+1: 
I was thinking of using it to run two or more applications that allocate memory on the GPU. Can the mallocMC state be saved/restored in the system memory before/after a context switch?

**Comment by ax3l** [comment]
oh it's all @slizzered ;)

**Comment by slizzered** [comment]
@anshumang I agree, that this would be a nice addition. Currently, saving/restoring the state is not directly implemented in mallocMC.

But all of the state is included in one chunk of allocated memory (the memPool that contains the heap), so it is possible to implement it. We just need to copy the whole memPool to host memory (#86  will be helpful). To recreate the memPool later, it must be again allocated and the contents copied from host.

Usually, mallocMC uses an implicit global object as the heap (which is hidden through macros). So we need to set the pointer to the object `mallocMCGlobalObject` as well. 
Alternatively, one can use the heap object directly (and use different object names directly in code)
```c++
int* a = (int*) mallocMC::mallocMCGlobalObject1.alloc(100 * sizeof(int));
int* b = (int*) mallocMC::mallocMCGlobalObject2.alloc(200 * sizeof(int));
```

**Comment by anshumang** [comment]
@slizzered Thanks for your reply ! I have a followup question.
CE > But all of the state is included in one chunk of allocated memory (the memPool that contains the heap),so it is possible to implement it. We just need to copy the whole memPool to host memory.
AG > I am thinking if we could selectively copy out only updated objects.

CE> To recreate the memPool later, it must be again allocated and the contents copied from host.
AG > All the memory objects should just be there when the context comes back, is it not? Faults (and copy back) should only be needed when oversubscribing in excess of available device memory. Please correct me if my assumption is wrong.

**Comment by slizzered** [comment]
@anshumang Oh I think I misunderstood your initial question. I thought it was about using a completely different heap and memory pool.

If I understand you now correctly, you want something like copy-on-write snapshots of the memory that can be rolled back. Unfortunately, this is not entirely possible with the level of abstraction that mallocMC provides right now:
mallocMC returns a raw pointer to memory. Until the memory of this pointer is freed again, mallocMC does not interact (or know about) with the access to the memory that is pointed to. So it can not see any writes that might happen. As a result, memory could be simply overwritten in the new context without any chance to intervene.

So to answer your follow-up questions:
AG > I am thinking if we could selectively copy out only updated objects.
CE > Right now I am not aware of a way to find out that memory was updated (there is no object abstraction within mallocMC and we don't have access to the CUDA runtime, which probably has such information)

AG > All the memory objects should just be there when the context comes back, is it not? Faults (and copy back) should only be needed when oversubscribing in excess of available device memory. Please correct me if my assumption is wrong.
CE > You are correct, but since it is not possible to selectively copy objects, you would need to duplicate the whole memoryPool. Depending on your application this might be too much to fit into the GPU memory, so you would probably have to backup the old memoryPool to system memory and delete it from the GPU (and restore it later).

**Comment by anshumang** [comment]
@sudkannan Adding Sudarsun (interested in checkpoint/restarts) in the thread...

**Comment by ax3l** [comment]
@anshumang just a side note: I am not sure if you are actually interested in [checkpoint handling in PIConGPU](https://github.com/ComputationalRadiationPhysics/picongpu/wiki/Develop-Plugins) which is realized via plugins.

`mallocMC` is just the memory manager for particle data (since massively parallel (de-)allocations of small chunks is a task of it's own).

**Comment by psychocoderHPC** [comment]
Copy a mallocMC buffer to host and later on back is only valid if the start pointer of the heap before and after the copy is the same. The reason is that we don't know how the allocator handle internel allocated chunks. If the allocator uses a linked list and store pointers to the chunks than the heap is only valid if all heap data are on the same adresses after the temporary copy to host.
